pretyping/evarsolve.ml

```diff
@@ -73,25 +72,33 @@ let refresh_universes ?(status=univ_rigid) ?(onlyalg=false) ?(refreshset=false)
        refresh_sort status ~direction s
       | _ -> t
       end
-    | Prod (na,u,v) -> 
-      mkProd (na, u, refresh ~onlyalg status ~direction v)
+    | Prod (na,u,v) ->
+       let v' = refresh ~onlyalg status ~direction v in
+       if v' == v then t else mkProd (na, u, v')
     | _ -> t
   (** Refresh the types of evars under template polymorphic references *)
   and refresh_term_evars onevars top t =
     match EConstr.kind !evdref t with
     | App (f, args) when is_template_polymorphic env !evdref f ->
       let pos = get_polymorphic_positions !evdref f in
-	refresh_polymorphic_positions args pos
+        refresh_polymorphic_positions args pos; t
     | App (f, args) when top && isEvar !evdref f -> 
-      refresh_term_evars true false f; 
-      Array.iter (refresh_term_evars onevars false) args
+       let f' = refresh_term_evars true false f in
+       let args' = Array.map (refresh_term_evars onevars false) args in
+       if f' == f && args' == args then t
+       else mkApp (f', args')
     | Evar (ev, a) when onevars ->
       let evi = Evd.find !evdref ev in
-      let ty' = refresh ~onlyalg univ_flexible ~direction:true (EConstr.of_constr evi.evar_concl) in
-	if !modified then 
-	  evdref := Evd.add !evdref ev {evi with evar_concl = EConstr.Unsafe.to_constr ty'}
-	else ()
-    | _ -> EConstr.iter !evdref (refresh_term_evars onevars false) t
+      let ty = EConstr.of_constr evi.evar_concl in
+      let ty' = refresh ~onlyalg univ_flexible ~direction:true ty in
+      if ty == ty' then t
+      else (evdref := Evd.add !evdref ev {evi with evar_concl = EConstr.to_constr !evdref ty'}; t)
+    | Sort s ->
+       (match ESorts.kind !evdref s with
+        | Type u when not (Univ.Universe.is_levels u) ->
```

Comment:
I don't understand this. If I call `refresh_universes` with `~onlyalg:false`, I'd expect all universes to be refreshed here, not just algebraics. Am I missing something? I would expect instead to call `refresh` on the sort.

Reply:
~onlyalg:true means algebraics are refreshed _and_ local universe levels which are algebraic (as declared in the evd) gets their status changed to non-algebraic. This ensures that the refreshed type can be used as a term (it won't contain algebraics). ~onlyalg:false means all algebraics and non-local universes. In that case we want to ensure (for template polymorphism) that fresh universes are generated replacing the global universes, to avoid capturing them, which would break the feature. Refreshing does not need to be done on non-algebraic local universes.